kube/src/service/auth.rs

```diff
@@ -0,0 +1,191 @@
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+
+use futures::{ready, Future};
+use http::{header::AUTHORIZATION, Request};
+use hyper::Body;
+use pin_project::pin_project;
+use tower::{layer::Layer, BoxError, Service};
+
+use crate::{config::Authentication, Result};
+
+/// `Layer` to decorate the request with `Authorization` header.
+pub struct AuthLayer {
+    auth: Authentication,
+}
+
+impl AuthLayer {
+    pub(crate) fn new(auth: Authentication) -> Self {
+        Self { auth }
+    }
+}
+
+impl<S> Layer<S> for AuthLayer
+where
+    S: Service<Request<Body>>,
+{
+    type Service = AuthService<S>;
+
+    fn layer(&self, service: S) -> Self::Service {
+        AuthService {
+            auth: self.auth.clone(),
+            service,
+        }
+    }
+}
+
+pub struct AuthService<S>
+where
+    S: Service<Request<Body>>,
+{
+    auth: Authentication,
+    service: S,
+}
+
+impl<S> Service<Request<Body>> for AuthService<S>
+where
+    S: Service<Request<Body>> + Clone,
+    S::Error: Into<BoxError>,
+{
+    type Error = BoxError;
+    type Future = AuthFuture<S>;
+    type Response = S::Response;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.service.poll_ready(cx).map_err(Into::into)
+    }
+
+    fn call(&mut self, mut req: Request<Body>) -> Self::Future {
+        // Comment from `AsyncFilter`
+        // > In case the inner service has state that's driven to readiness and
+        // > not tracked by clones (such as `Buffer`), pass the version we have
+        // > already called `poll_ready` on into the future, and leave its clone
+        // > behind.
+        let service = self.service.clone();
+        let service = std::mem::replace(&mut self.service, service);
+
+        let auth = self.auth.clone();
+        let request = async move {
+            // If using authorization header, attach the updated value.
+            auth.to_header().await.map_err(BoxError::from).map(|opt| {
+                if let Some(value) = opt {
+                    req.headers_mut().insert(AUTHORIZATION, value);
+                }
+                req
+            })
+        };
+
+        AuthFuture {
+            state: State::Request(Box::pin(request)),
+            service,
+        }
+    }
+}
+
+#[pin_project(project = StateProj)]
+#[derive(Debug)]
+enum State<F, G> {
+    /// Waiting for the request future
+    Request(#[pin] F),
+    /// Waiting for the response future
+    Response(#[pin] G),
+}
```

Comment:
You got pin_project working quickly :-)
Implementation and naming is readable and looks sensible. Similar to [AsyncResponseFuture](https://github.com/tower-rs/tower/blob/2a7d47adda559764c7b4ba8dc64b44ead0385def/tower/src/filter/future.rs#L17-L89) AFAIKT, presumably based on that so we can have better names?

Reply:
Yeah, I did it based on what I saw in `tower`. I don't understand them fully, but it compiled and seems to work :)